### ISSUE TITLE
Integrate 2 factor authentication (2FA) (Steam Guard)

### ISSUE THREAD

**Issue by webbertakken** [game-ci/steam-deploy#4]
**Is your feature request related to a problem? Please describe.**
Currently anyone using a Steam account that has Steam Guard 2-factor authentication enabled will automatically fail their workflow, because the email will arrive, but there is currently no way to provide the code to the action.

**Describe the solution you'd like**

1. Using dedicated API keys.

**Verdict:** ❌ Not supported in current Steamworks SDK

2. Removing steam guard (2FA) from builder account

**Verdict:** ❌ I asked both Steam Support and Steam Partner Support, and both came with the same answer: _"Not being able to disable Steam Guard email level protection is Steams "deliberate design choice" and as such policy for any steam partner account"_

3. Following Steams advice by manually interacting with the workflow in order to authenticate each unique machine once.

**Verdict:** ✔ This could work! Since Github's machine signature seems to be very reliant, we would probably only have to authenticate Github Actoins once (per project per steam account).

**Additional context**

The reply from Steam:

> For security reasons, we don't allow users in partner accounts to disable Steam Guard.
> 
> If your uploads are all being issued from a single machine (or virtual machine), you'll want to interactively log in with `steamcmd` one time and issue the `set_steam_guard_code` command with the code you were emailed. You can test immediately by issuing a `logon` and ensuring that it doesn't ask for a code. If it logs on, that machine should be able to log on in the future without having to resubmit a Steam Guard code.
> 
> You can also add an optional `Steam Guard Code` parameter that, if present, calls `set_steam_guard_code` prior to logging in and executing the build/depot scripts. That way, if `steamcmd` ever requires a Steam Guard code again (which can happen after an update), it will fail once, but you can just run it again with a Steam Guard code parameter and you won't need to log in interactively again.
> 
> Let me know if you have any other questions!
> 
> Best regards,
> ~Redacted~

I would say that in practise this should be very achievable.

Related: #3

**Comment by Denaton** [comment]
@webbertakken Will that work since its from different IP-addresses each time?

Edit 1; Could a ugly solution be to have a build email that read the code and place it in the steamcmd when the code is requested?
Not optimal but still something..

Edit 2; Done some more reading, seems that 2FA generates `maFiles` that that one is only needed to be generated once and then we can use it as many times as we want. I have seen a few other projects that has successfully generated `maFiles` but they all needed confirmation before generating, either by SMS or Email.

**Comment by Jedimaster4559** [comment]
It seems that some setup with cashing the 2FA files is probably the best/most manageable option here. One thing that was very helpful as I was debugging this for myself was this website here: [Steam Error Codes](https://steamerrors.com/)<br><br>

Another thing I just found in the steam documentation:

> If you are using steamcmd from a machine or VM that gets re-imaged frequently, you should include the sentry file in your image so you won't be prompted for a Steam Guard every time. The sentry file is located in `<Steam>\ssfn<hashOfNumbers>`

I still need to go poke around and find this in linux, but it seems this may also be a resolution to this problem.

**Comment by Jedimaster4559** [comment]
So, I was able to get this working in my own setup. I did something kinda similar to what I was mentioning above. The Sentry file does maintain the credentials so you won't have to constantly re-authenticate anytime your pipeline runs.

I found this file in the /home/steam/Steam/ directory. I was able to just upload it to my repo and copy it in to the directory to get things working. I imagine there are many better approaches to making that persistent, but this ended up working for me.

Something I didn't try, but that might work is logging in on a separate machine and copying the file from there. That could simplify things some, however I would guess that is less likely to work.

**Comment by danmastrow** [comment]
Similiar to what @Jedimaster4559 said, I think the only approach might be to use the Sentry(ssfn) + Config files.

My current 'manual' approach in order for the Github agent to use steamcmd with SteamGuard enabled was to:

1. Set _${{ secrets.STEAM_PASSWORD }}_ and _${{ secrets.STEAM_USERNAME }}_ as Github secrets for the repository.

2. Explicitly DO NOT SET _${{secrets.STEAM_MFA }}_ for now, we want an empty value, and we are expecting the login to fail.

3. Add step
```
 - name: Run deploy
        run: steamcmd +login ${{ secrets.STEAM_USERNAME }} ${{ secrets.STEAM_PASSWORD }} ${{secrets.STEAM_MFA }}
```

This command will fail, but will trigger sending an MFA code to the email associated with the _${{ secrets.STEAM_PASSWORD }}_

4. Set the MFA code received by email to ${{secrets.STEAM_MFA }}

5. Add an export artifact step to your job, so that you can retrieve the config.vdf and Sentry file 
```
      # Output Steam artifacts created by a successful login. 
       # The only ones we care about are the /config/config.vdf and the ssfnXXXXXXX AKA Sentry file.
       
      - name: Export artifact
        uses: actions/upload-artifact@v2
        with:
          name: Steam
         path: '/home/runner/Steam/'

```
6. Run your pipeline again (with the MFA Secret set)
This time your steamcmd login should be successful, as you are using a valid MFA code, and also will export the /Steam/ folder.
NOTE: This is not the end of the setup, this will only work once, but is needed to retrieve the config + ssfn files.

7. Manually checkin the config.vdf and ssfn file into your repo

8. Add more steps to your pipeline to copy the config.vdf and ssfn files into the expected places on the agent, for every run
```
      - name: Create config folder
        run: mkdir -p /home/runner/Steam/config

      - name: Copy config
        run: mv -v config.vdf /home/runner/Steam/config/config.vdf
        
      - name: Copy sentry file
        run: mv -v ssfn* /home/runner/Steam/
```

9. Delete the ${{secrets.STEAM_MFA }} secret.
10. Remove the Export artifact step (no longer needed)


I would recommend improving this to avoid checking in the ssfn and config files.
One way to do this would be to put them into SECRETS.
To make an even better user experience for this library you could instead automatically create the secrets for the user, and avoid the Export Artifact step
https://github.com/marketplace/actions/create-github-secret-action

The 'One time' manual process of using a real MFA code is likely unavoidable, but it can be similiar to the License setup for the unity build actions.

**Comment by Jedimaster4559** [comment]
I agree, they should be put into the secrets section for security reasons. @danmastrow's explanation of how this works in a more practical way is really solid. I also like to run the automation on it's own steam account that only has permissions to do the uploads to steam. I've found that's better than using my own personal credentials anywhere.

Also, in case someone happens to find this while googling. If you have a steam developer account, it is not possible to disable two-factor authentication. If you want to do some kind of automation of your steam upload process, you will have to find a solution for the two factor, whether that is a manual step, or is something more like what you see here.

**Comment by webbertakken** [comment]
Perhaps we could run the docker image locally with some parameter (or interactive) to acquire the artifact.
Then we could set the contents of the vdf file as a secret in GitHub. 

That way you wouldn't have to adapt your workflow 3 times.

**Comment by webbertakken** [comment]
This should now technically work after #8. Would be nice if someone could confirm this.

**Comment by MCOfficer** [comment]
Sorry, how is one supposed to use #8 ? also, would using https://github.com/marketplace/actions/steam-totp be an option?

**Comment by webbertakken** [comment]
> Sorry, how is one supposed to use #8 ? also, would using https://github.com/marketplace/actions/steam-totp be an option?

You only have to authorise a machine once, so you can just add it to your secrets after you get the first email.

This action doesn't really care about how to retrieve the two-factor authentication code, so you can probably use the action you've indicated. Have you tried it?

**Comment by MCOfficer** [comment]
> > Sorry, how is one supposed to use #8 ? also, would using [github.com/marketplace/actions/steam-totp](https://github.com/marketplace/actions/steam-totp) be an option?
> 
> You only have to authorise a machine once, so you can just add it to your secrets after you get the first email.
> 
> This action doesn't really care about how to retrieve the two-factor authentication code, so you can probably use the action you've indicated. Have you tried it?

I haven't, and since i don't (yet?) have access to a steamworks application and don't want to pay 100$ for testing, i won't in the immediate future. I was just doing research, but when i do try it, i'll let you know the results.

**Comment by hookkshot** [comment]
I have tried this, many times the steam_cmd will tell me I need to reauthenticate. It works ok. But we need something a bit more reliable if we can get something together. Steam really need proper API access for this stuff.

**Comment by tom-leys** [comment]
@webbertakken Continuing our discussion from https://github.com/game-ci/steam-deploy/pull/13

I thought I had this working for a short time, but I ran into a couple of major issues. 

1. /steam-deploy does not ingest ssfn files
2. It also does not appear to write them out.

I also considered github.com/marketplace/actions/steam-totp but you are not supposed to know your shared secret. The only way to get it is to install some highly suspect tools (full of trojans most likely). I opted out of that method.

Here is what seems to show promise for me

1. Install [valve's offical steamcmd](https://partner.steamgames.com/doc/sdk/uploading#1) on your local machine
2. Log in using your build account on your desktop (I did it in windows)
3. Answer the challenge requiring you to read your email.
4. quit after a good login. (type quit)
5. If you type `steamcmd +login <user> <password> +quit` on the console it should now login OK on your desktop and quit again.
6. Observe the folder which contained steamcmd before. It now contains the `ssfn...` file and the config/config.vdf file.
7. Put these in your repository. I [encrypted the ssfn file with another secret](https://docs.github.com/en/actions/security-guides/encrypted-secrets#limits-for-secrets) but that part is optional.

NB: If you encrypt, the command on your PC is `gpg --symmetric --cipher-algo AES256 ssfn_xxx` (you will need to replace _xxx). You can use the md5sum calculated in the action below to confirm the decryption results in the same file you have.

Now in your GitHub action

1. Install steamcmd
2. restore this config and ssfn file
3. Test the login works using steamcmd
4. Note that /steam-deploy/ still fails to log in

As of this writing I'm configuring a manual upload using steamcmd directly

```yml
# yaml-language-server: $schema=https://json.schemastore.org/github-workflow.json
# Use this to validate: https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml
name: godot-export

# To configure On, see : https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows
on: push
  # push:
  #   branches: [ $default-branch ]
  # pull_request:
  #   branches: [ $default-branch ]

env:
  CARGO_TERM_COLOR: always
  CACHE_ON_FAILURE: true

jobs:

  steam-upload:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v2

      - name: Create config folder
        run: mkdir -p /home/runner/Steam/config

      - name: Decrypt steam secrets
        run: |
          chmod ug+x ${{github.workspace}}/steam/decrypt_secret.sh
          ${{github.workspace}}/steam/decrypt_secret.sh
        env:
          working-directory: ${{github.workspace}}/steam
          LARGE_SECRET_PASSPHRASE: ${{ secrets.STEAM_DECRYPT }}
          DECRYPT_FILE: ${{github.workspace}}/steam/ssfn.gpg
          FILE_DEST: /home/runner/Steam/ssfn_xxxx

      # Note, you will have a different ssfn name than me
      - name: ssfn md5sum
        run: md5sum /home/runner/Steam/ssfn_xxxx

      - name: Copy config
        run: cp -v ${{github.workspace}}/steam/config/config.vdf /home/runner/Steam/config/config.vdf
        
      - name: Setup steamcmd
        uses: CyberAndrii/setup-steamcmd@e19cd1516315ce46dbfffa47193f92fe42d1419e
        
      # you can see from the output here that the login is OK
      - run: steamcmd +login ${{ secrets.STEAM_USERNAME }} ${{ secrets.STEAM_PASSWORD }} +quit

# This fails to log in
#       - name: Deploy to Steam ☁
# #        if: always()
#         uses: game-ci/steam-deploy@b2d01067dacd4efc7c4c4722dccfe05d5fff9638
#         with:
#           username: ${{ secrets.STEAM_USERNAME }}
#           password: ${{ secrets.STEAM_PASSWORD }}
#           appId: 1234
#           buildDescription: tag-${{ github.ref }}-${{github.sha}} 
#           rootPath: ${{github.workspace}}/game/build/
#           depot1Path: windows
#           depot2Path: linux
#           depot3Path: macos
#           releaseBranch: PreRelease
#   #        localContentServer: LocalContentServer
   
      # Output Steam artifacts created by a successful login. 
      # The only ones we care about are the /config/config.vdf and the ssfnXXXXXXX AKA Sentry file.
       
      - name: Display steam config file listing
        if: always()
        run: ls -la -R
        working-directory: /home/runner/Steam/

      # Note: This file contains secrets, don't leave it in production.
      - name: Export artifact
        if: always()
        uses: actions/upload-artifact@v2
        with:
          name: Steam
          path: '/home/runner/Steam/'
```

Decrypt secret (`decrypt_secret.sh`) exists only to hide the arguments passed I believe:

```sh
#!/bin/sh

# Decrypt the file
# --batch to prevent interactive command
# --yes to assume "yes" for questions
gpg --quiet --batch --yes --decrypt --passphrase="$LARGE_SECRET_PASSPHRASE" \
--output $FILE_DEST $DECRYPT_FILE
```

**Comment by tom-leys** [comment]
@webbertakken (or others) feel free to [join my discord](https://discord.gg/tRCuSNH) for realtime discussion of this or similar issues. You'll see I'm automating a Godot build process.

**Comment by webbertakken** [comment]
Note that GameCI already has a [community discord](https://game.ci/discord).

**Comment by tom-leys** [comment]
My GitHub actions have been working as per the above instructions for a few weeks now no problems.

**Comment by mikeage** [comment]
~~This might be a steamcmd specific problem, but I'm having trouble getting this working -- it seems like even after backing up the ssfn* and config/config.vdf files, I'm always being prompted for MFA.~~

~~Is there a difference between using a dedicated builder account (which I plan to do) and a personal account with SteamGuard? I'm currently using the latter~~

Edit: it seems that after saving the ssfn/config, I needed to delete the secrets for both the MFA and the password, not just the MFA. 

Edit again: it seems like mobile steam guard might be problematic, but with an email OTP user, it's working just fine. Well, not the action -- see below

That said, it doesn't work with the existing action, because it doesn't seem to be able to use these files. @tom-leys , do you have a full example including your upload? Alternatively, @webbertakken , is there any way to not require a new MFA on each build -- despite the documentation update of #13 , it doesn't seem like this is actually working -- am I missing something?

**Comment by davidmfinol** [comment]
@tom-leys @mikeage 
Have either of you seen this error or happen to know what this error means?: 
```
Loading Steam API...KeyValues Error: LoadFromBuffer: missing {   (current key: '***') in file *** [offset: 19]

src/tier1/KeyValues.cpp (2967) : Assertion Failed: Error while parsing text KeyValues for resource ***
KeyValues Error: LoadFromBuffer: missing {   (current key: '***') in file *** [offset: 19]

src/tier1/KeyValues.cpp (2967) : Assertion Failed: Error while parsing text KeyValues for resource ***
OK
Logging in user '***' to Steam Public...
password: KeyValues Error: LoadFromBuffer: missing {   (current key: '***') in file *** [offset: 19]

src/tier1/KeyValues.cpp (2967) : Assertion Failed: Error while parsing text KeyValues for resource ***
FAILED (Invalid Password)
```
I get this error when I run `steamcmd +login "$username"   +run_app_build $(pwd)/manifest.vdf +quit` after copying the config.vdf and ssfn files to their appropriate location.

**Comment by mikeage** [comment]
Sorry, I have not

**Comment by davidmfinol** [comment]
Thanks for replying @mikeage 
I think I was able to get past that issue, but now I'm seeing the issue where I can successfully authenticate to Steam and upload builds, but the depots that are uploaded are all empty.
Any chance you've seen something like that?

**Comment by davidmfinol** [comment]
I've used the information here to update this action to correctly use the ssfn* files. Thanks everybody, especially @mikeage and @tom-leys , for the input here.
Note that using Steam totp no longer works with this action, but I'll create another issue for that.

**Comment by MCOfficer** [comment]
TOTP follow up is #17
